pallets/messages/src/lib.rs

```diff
@@ -337,27 +362,81 @@ impl<T: Config> Pallet<T> {
 			.map_err(|_| Error::<T>::InvalidMessageSourceAccount)?)
 	}
 
-	/// Gets a messages for a given schema-id and block-number.
+	/// Retrieve the messages for a particular intent and block range (paginated)
+	pub fn get_messages_by_intent_id(
+		intent_id: IntentId,
+		pagination: BlockPaginationRequest,
+	) -> BlockPaginationResponse<MessageResponseV2> {
+		let mut response = BlockPaginationResponse::new();
+
+		// Request Validation
+		if !pagination.validate() {
+			return response
+		}
+
+		// Schema Fetch and Check
+		let intent = match <T>::SchemaProvider::get_intent_by_id(intent_id) {
+			Some(intent) => intent,
+			None => return response,
+		};
+
+		let mut from_index: u32 = pagination.from_index;
+
+		let _ = (pagination.from_block..pagination.to_block).try_for_each(|block_number| {
+			let list: Vec<MessageResponseV2> = Self::get_messages_by_intent_and_block(
+				intent_id,
+				intent.payload_location,
+				block_number.into(),
+			);
+
+			// Max messages in a block are constrained to MessageIndex (u16) by the storage,
+			// so this is a safe type coercion. Just to be safe, we'll trap in in debug builds
+			let list_size: u32 = list.len() as u32;
```

Comment:
Nit: I think we can avoids casting before asserting correctness. This also improves readability.
```
let list_size = list.len();
debug_assert!(list_size <= u16::MAX as usize, '....");
let list_size = list_size as u32;
```